### PR TITLE
Replace setInventoryKnown with a rolling bloom filter.

### PR DESCRIPTION
**Pull request by gmaxwell**
Mruset setInventoryKnown was reduced to a remarkably small 1000
 entries as a side effect of sendbuffer size reductions in 2012.

This removes setInventoryKnown filtering from merkleBlock responses
 because false positives there are especially unattractive and
 also because I'm not sure if there aren't race conditions around
 the relay pool that would cause some transactions there to
 be suppressed. (Also, ProcessGetData was accessing
 setInventoryKnown without taking the required lock.)

### CODE REVIEW

**Comment by dcousens** [comment]
@gmaxwell with this change, `mruset` can be completely removed.

**Comment by pstratem** [comment]
@gmaxwell how large is the filter?

**Comment by sipa** [comment]
1.37 MIB per peer...

**Comment by pstratem** [comment]
@sipa in that case utACK

**Comment by gmaxwell** [comment]
hm. Is it that big? My math must be wrong, I was thinking it was more like 350k.

**Comment by sipa** [comment]
Rolling bloom filters are 4 times as large as normal ones.

**Comment by sipa** [comment]
Eh, messed up my math. This should result in 700 KiB per peer.

**Comment by sipa** [comment]
Confirmed with a sneaky `fprintf` call: 35 KiB for the address tables, and 702 KiB for the inventory tables.

**Comment by sipa** [comment]
utACK. Also, mruset can go away indeed.

**Comment by dcousens** [comment]
utACK

**Comment by laanwj** [comment]
700kiB per peer is quite a lot for a P2P application in itself, but more relevant is how it compares to `mruset` overhead when keeping the functionality at the same level. I suppose it's much better?

**Comment by sipa** [comment]
An mruset for the same size (last 50000 entries) would be almost 4 MiB. Of
course, they are only size 1000 currently...

**Comment by laanwj** [comment]
Yes, that's why I mentioned "functionally equivalent", comparing against the current state isn't fair.

**Comment by sipa** [comment]
I do agree 700 KiB is pretty high per peer.

Do we really need a 1/1000000 false positive rate? An FP will just cause us
to not relay a particular transaction to one peer, but we'll still relay it
to others (they have other nonces).

Also, a data structure exists that is 25% more efficient than the current
CRollingBloomFilter.

**Comment by pstratem** [comment]
@sipa Care to share? :)

**Comment by laanwj** [comment]
> Do we really need a 1/1000000 false positive rate? An FP will just cause us
to not relay a particular transaction to one peer, but we'll still relay it
to others (they have other nonces).

Ideally what we'd want here is the opposite of a bloom filter. False positives are bad, as they cause peers to miss out on transactions, but the occasional false negative is ok, as they might get it INVed twice.

**Comment by sipa** [comment]
Agree... unfortunately I don't know of any better way to do that than an mruset...

Anyway, #7113 reduces this to 526 KiB per peer.

File: src/net.cpp [hunk 1]
```diff
@@ -2370,6 +2370,7 @@ CNode::CNode(SOCKET hSocketIn, const CAddress& addrIn, const std::string& addrNa
     nSendOffset = 0;
     hashContinue = uint256();
     nStartingHeight = -1;
+    setInventoryKnown.reset();
```
Comment (jtimon) [inline src/net.cpp]: Won't this undo the setInventoryKnown(50000, 0.000001) initialization?

**Comment by jtimon** [comment]
utACK

File: src/net.cpp (hunk 1 above)
Comment (sipa) [inline src/net.cpp]: No, it only wipes the set, not the parameters.

**Comment by gmaxwell** [comment]
@sipa I'm not really excited with having false positives at all here; mostly due to the case where you only have one peer, a FP means you may not relay your own transaction.

In a universe where software complexity had no cost, I'd give MRUsets to (or even just the first few) outbound peers, and higher FP rolling bloom filters to all the others.

**Comment by sipa** [comment]
utACK

**Comment by pstratem** [comment]
needs rename of setInventoryKnown to filterInventoryKnown

won't a false positive here break block relaying?

**Comment by sipa** [comment]
Yes very good point, though I think that after #6494 we may be able to get
rid of using inventory for blocks entirely.

**Comment by pstratem** [comment]
there doesn't seem to be a strong reason to filter block inv messages

i say just send them without filtering

**Comment by pstratem** [comment]
ACK

**Comment by sipa** [comment]
Updated #7125 to not use setInventoryKnown for blocks anymore.

File: src/net.h [hunk 2]
```diff
@@ -492,8 +491,9 @@ class CNode
     {
         {
             LOCK(cs_inventory);
-            if (!setInventoryKnown.count(inv))
-                vInventoryToSend.push_back(inv);
+            if (inv.type == MSG_TX && filterInventoryKnown.contains(inv.hash))
```
Comment (sipa) [inline src/net.h]: This test is not enough. SendMessages checks the filter again.

**Comment by sipa** [comment]
Needs rebase on top of #7129.

**Comment by sipa** [comment]
See #7133.

**Comment by gmaxwell** [comment]
replaced by #7133.

**Pull request closed without merging**